File: le-serviceflows/core/src/main/resources/com/latticeengines/serviceflows/core/transforms/std_visidb_ds_title_level.py

#
# $LastChangedBy$
# $LastChangedDate$
# $Rev$
#

import re

def metadata():
    return {  'ApprovedUsage'   : 'ModelAndModelInsights'
            , 'Description'     : 'Numeric score corresponding to Job Title level; Senior Titles have higher values'
            , 'DisplayDiscretizationStrategy' : '{"linear": { "minValue":0,"stepSize":1,"minSamples":100,"minFreq":0.01,"maxBuckets":5,"maxPercentile":1}}'
            , 'DisplayName'     : 'Job Title Seniority'
            , 'Category'        : 'Lead Information'
            , 'StatisticalType' : 'ordinal'
            , 'Tags'            : 'Internal' }



def ds_title_issenior(n):
    if re.search('(?<!\w)sr(?!\w)|senior', n):
        return 1
    return 0

def ds_title_ismanager(n):
    if re.search('(?<!\w)mgr(?!\w)|manag', n):
        return 1
    return 0

def ds_title_isdirector(n):
    if re.search('(?<!\w)dir', n):
        return 1
    return 0

def ds_title_isvpabove(n):
    if re.search('(?<!\w)vp(?!\w)|pres|chief|(?<!\w)c\wo(?!\w)', n):
        return 1
    return 0


def std_visidb_ds_title_level(n):
    if n is None:
        return None

    n = n.lower()

    return ds_title_issenior(n) + (2*ds_title_ismanager(n) + (4*ds_title_isdirector(n) + 8*ds_title_isvpabove(n)))


def transform(args, record):
    column = args["column"]
    value = record[column]
    return std_visidb_ds_title_level(value)

File: le-transform/src/main/java/com/latticeengines/transform/v2_0_25/functions/StdVisidbDsTitleLevel.java
package com.latticeengines.transform.v2_0_25.functions;

import java.util.Map;
import java.util.regex.Pattern;

import com.latticeengines.transform.exposed.RealTimeTransform;
import com.latticeengines.transform.exposed.metadata.ApprovedUsage;
import com.latticeengines.transform.exposed.metadata.Category;
import com.latticeengines.transform.exposed.metadata.FundamentalType;
import com.latticeengines.transform.exposed.metadata.StatisticalType;
import com.latticeengines.transform.exposed.metadata.Tag;
import com.latticeengines.transform.exposed.metadata.TransformMetadata;

public class StdVisidbDsTitleLevel implements RealTimeTransform {

    private static final long serialVersionUID = 5331340048981069485L;

    public StdVisidbDsTitleLevel() {
    }

    public StdVisidbDsTitleLevel(String modelPath) {
    }

    public Object transform(Map<String, Object> arguments, Map<String, Object> record) {
        String column = (String) arguments.get("column");
        Object n = record.get(column);

        if (n == null)
            return 0;

        String s = n.toString().toLowerCase();

        return calculateStdVisidbDsTitleLevel(s);
    }

    public static int calculateStdVisidbDsTitleLevel(String s) {
        return StdVisidbDsTitleLevel.dsTitleIssenior(s)
                + (2 * StdVisidbDsTitleLevel.dsTitleIsmanager(s) + (4 * StdVisidbDsTitleLevel.dsTitleIsdirector(s) + 8 * StdVisidbDsTitleLevel
                        .dsTitleIsvpabove(s)));
    }

    private static int dsTitleIssenior(String s) {
        return Pattern.matches("(.*?\\b)sr(\\b.*?)|(.*?\\b)senior(\\b.*?)", s) ? 1 : 0;
    }

    private static int dsTitleIsmanager(String s) {
        return Pattern.matches("(.*?\\b)mgr(.*)|(.*)manag(.*)", s) ? 1 : 0;
    }

    private static int dsTitleIsdirector(String s) {
        return Pattern.matches("(.*?\\b)dir(.*)", s) ? 1 : 0;
    }

    private static int dsTitleIsvpabove(String s) {
        return Pattern.matches(
                "(.*?\\b)vp(\\b.*?)|(.*?\\b)pres(.*)|(.*?\\b)chief(\\b.*?)|(.*)(?<!\\w)c[tefmxdosi]o(?!\\w)(.*)", s) ? 1
                : 0;
    }

    @Override
    public TransformMetadata getMetadata() {
        TransformMetadata metadata = new TransformMetadata();
        metadata.setApprovedUsage(ApprovedUsage.MODEL_MODELINSIGHTS);
        metadata.setCategory(Category.LEAD_INFORMATION);
        metadata.setFundamentalType(FundamentalType.NUMERIC);
        metadata.setStatisticalType(StatisticalType.ORDINAL);
        metadata.setTags(Tag.INTERNAL_TRANSFORM);
        metadata.setDisplayDiscretizationStrategy("{\"linear\": { \"minValue\":0,\"stepSize\":1,\"minSamples\":100," //
                + "\"minFreq\":0.01,\"maxBuckets\":5,\"maxPercentile\":1}}");
        metadata.setDescription("Numeric score corresponding to Job Title level; Senior Titles have higher values");
        metadata.setDisplayName("Job Title Seniority");
        return metadata;
    }
}
